src/psyclone/tests/psyir/transformations/transformations_test.py

```diff
@@ -444,6 +444,99 @@ def create_loops(body):
     omplooptrans.validate(loops[0])
 
 
+def test_omplooptrans_apply_firstprivate(fortran_reader, fortran_writer):
+    ''' Test applying the OMPLoopTrans in cases where a firstprivate
+    clause it is needed to generate functionally equivalent code than
+    the starting serial version. '''
+
+    # Example with a conditional write and a OMPParallelDoDirective
+    psyir = fortran_reader.psyir_from_source(f'''
+        subroutine my_subroutine()
+            integer :: ji, jj, jk, jpkm1, jpjm1, jpim1, scalar1, scalar2
+            real, dimension(10, 10, 10) :: zwt, zwd, zwi, zws
+            do jk = 2, jpkm1, 1
+              do jj = 2, jpjm1, 1
+                do ji = 2, jpim1, 1
+                   if (.true.) then
+                      scalar1 = zwt(ji,jj,jk)
+                   endif
+                   scalar2 = scalar1 + zwt(ji,jj,jk)
+                   zws(ji,jj,jk) = scalar2
+                enddo
+              enddo
+            enddo
+        end subroutine''')
+    omplooptrans = OMPParallelLoopTrans()
+    loop = psyir.walk(Loop)[0]
+    omplooptrans.apply(loop)
+    expected = '''\
+!$omp parallel do default(shared), private(ji,jj,jk,scalar2), \
+firstprivate(scalar1), schedule(auto)
+do jk = 2, jpkm1, 1
+  do jj = 2, jpjm1, 1
+    do ji = 2, jpim1, 1
+      if (.true.) then
+        scalar1 = zwt(ji,jj,jk)
+      end if
+      scalar2 = scalar1 + zwt(ji,jj,jk)
+      zws(ji,jj,jk) = scalar2
+    enddo
+  enddo
+enddo
+!$omp end parallel do\n'''
+
+    gen = fortran_writer(psyir.children[0].children[0])
+    assert expected == gen
+
+
+@pytest.mark.xfail(reason="Issue #598: Firstprivate still misses some cases")
+def test_omplooptrans_apply_firstprivate_fail(fortran_reader, fortran_writer):
+    ''' Test applying the OMPLoopTrans in cases where a firstprivate
+    clause it is needed to generate functionally equivalent code than
+    the starting serial version.
+
+    This examples are more complex than the previous test and the
+    transformation still fails to get them correctly.
+    '''
+
+    # Example with a read before write and a OMPParallelDirective
+    psyir = fortran_reader.psyir_from_source('''
+        subroutine my_subroutine()
+            integer :: ji, jj, jk, jpkm1, jpjm1, jpim1, scalar1, scalar2
+            real, dimension(10, 10, 10) :: zwt, zwd, zwi, zws
+            do jk = 2, jpkm1, 1
+              do jj = 2, jpjm1, 1
+                do ji = 2, jpim1, 1
+                   scalar2 = scalar1 + zwt(ji,jj,jk)
+                   scalar1 = 3
+                   zws(ji,jj,jk) = scalar2 + scalar1
+                enddo
+              enddo
+            enddo
+        end subroutine''')
+    omplooptrans = OMPParallelLoopTrans()
+    loop = psyir.walk(Loop)[0]
+    omplooptrans.apply(loop)
+    expected = '''\
+!$omp parallel do default(shared), private(ji,jj,jk,scalar2), \
+firstprivate(scalar1), schedule(auto)
+do jk = 2, jpkm1, 1
+  do jj = 2, jpjm1, 1
+    do ji = 2, jpim1, 1
+      if (.true.) then
```

Comment:
In fact, this test will already fail because this line doesn't correspond to the input source ;-)

Reply:
Ups, now it is properly checked in the previous test thanks to the force option.